File: Java/Problem4.java
import java.util.Locale;

class Solution {
    public String shortestPalindrome(String s) {
        s = s.toLowerCase();  // assign the string in lowercase
        int j = s.length() - 1;
        String newCharacter = "";
        int sameCharacter = 0;

        while (sameCharacter != (s.length() / 2)) {
            j = s.length() - 1;
            sameCharacter = 0;

            for (int i = 0; i < s.length() / 2; i++) {
                if (s.charAt(i) == s.charAt(i + j)) {
                    j -= 2;
                    sameCharacter++;
                } else {
                    newCharacter = String.valueOf(s.charAt(i + j));
                    for (char c : newCharacter.toCharArray()) {
                        s = s.substring(0, sameCharacter) + c + s.substring(sameCharacter);
                        sameCharacter++;
                    }
                    break;
                }
            }
        }
        return s;
    }
}

File: TypeScript/Problem4.ts
function shortestPalindrome(s: string): string {
    s = s.toLowerCase();  // assign the string in lowercase
    let j = s.length - 1;
    let new_character = "";
    let same_character = 0;

    while (same_character !== Math.floor(s.length / 2)) {
        let i = 0;
        j = s.length - 1;
        same_character = 0;

        for (i = 0; i < Math.floor(s.length / 2); i++) {
            if (s[i] === s[i + j]) {
                j -= 2;
                same_character += 1;
            } else {
                new_character = s[i + j];
                for (const c of new_character) {
                    s = s.slice(0, same_character) + c + s.slice(same_character);
                    same_character += 1;
                }
                break;
            }
        }
    }
    return s;
}

